include/ensmallen_bits/lbfgs/lbfgs_impl.hpp

```diff
@@ -86,11 +86,17 @@ double L_BFGS::ChooseScalingFactor(const size_t iterationNum,
     // Get s and y matrices once instead of multiple times.
     const arma::Mat<CubeElemType>& sMat = s.slice(previousPos);
     const arma::Mat<CubeElemType>& yMat = y.slice(previousPos);
-    scalingFactor = dot(sMat, yMat) / dot(yMat, yMat);
+    
+    const CubeElemType tmp   = arma::dot(yMat, yMat);
+    const CubeElemType denom = (tmp != CubeElemType(0)) ? tmp : CubeElemType(1);
```

Comment:
Why divide by 1 here and not some small number that won't cause overflow?  I would worry that this strategy would not capture the intuition of setting the scaling factor to be large when `dot(yMat, yMat)` is small.

Reply:
One problem is that in relation to the numerator `dot(sMat, yMat)`, we don't know which "small" number to use as the denominator, or whether a "small" number would be appropriate at all.

If the denominator is 0, this would imply that we're in all sorts of trouble with the provided data.

Given the lack of sufficient information, the best "non-informative" choice to avoid the generation of NaNs seems to be the mathematical equivalent of a no-op in the division, ie. denominator = 1.

This has parallels with using 1 on the diagonals of covariance matrices when encountering problematic data.  Occam's razor may also be applicable here.

Reply:
Yes, I see what you mean now.  I think there's a case to be made to just take `scalingFactor` as "some large value that's not quite inf", but that case is easier to poke holes in than yours (e.g. if `dot(sMat, yMat)` is actually smaller than `dot(yMat, yMat)`, but both are represented in floating point as 0).  So, I agree with your reasoning.